src/framework/components/esmscript/system.js

```diff
@@ -0,0 +1,133 @@
+import { Debug } from '../../../core/debug.js';
+import { ComponentSystem } from '../system.js';
+import { EsmScriptComponent } from './component.js';
+import { EsmScriptComponentData } from './data.js';
+import { DynamicImport } from '../../handlers/esmscript.js';
+
+/**
+ * Allows scripts to be attached to an Entity and executed.
+ *
+ * @augments ComponentSystem
+ */
+class EsmScriptComponentSystem extends ComponentSystem {
+    /**
+     * Create a new EsmScriptComponentSystem.
+     *
+     * @param {import('../../app-base.js').AppBase} app - The application.
+     * @hideconstructor
+     */
+
+    _components = new Set();
```

Comment:
From performance point of view, we loop over this set a lot, and I wonder if we need to convert it to an array using a dirty flag, and use for loop on it insead.

Reply:
I think the forEach is not idea and can be swapped out with a for of. Is there something inherently slow with a Set? I've run some tests and seems comparable with an array, the forEach iteration is much slower tho

Reply:
From my tests few months ago, looping over Set is maybe 3x slower than for(..) loop over an array. But yeah, hard to know it this would be worth in this case. For large project with 1000s of script instances maybe, but probably not even in this case.

Reply:
I will update using a for of loop which is definitely faster than the current forEach

Reply:
Even for of loop produces an iterator from a set, while for(;;) loops with just array.length is faster.
In case of rare modifications to array, it is beneficial to prioritise fast iteration through array than ease and speed of modification of a set.

Reply:
Yeah, for storing of mesh instancing is layers and similar I keep both the set and an array, one for fast lookup, one for fast iteration. Not great, but works. But it's hard to really confirm this is beneficial. For each loop over the meshes, we spend tons of time typically doing operations on them, (culling, rendering ...) so the win is relatively very small, perhaps even non-existing. But maybe it all adds up to something.

Reply:
The thinking was it's reasonable to assume the set could be modified during iteration, so mainly for safety, but also I get wildly different perf results with different browsers, and it seems to change with each release https://www.measurethat.net/Benchmarks/Show/28356/0/iterate-set-vs-arr